File: extensions/ip-rotator/src/main/java/lavaplayer/extensions/iprotator/YoutubeIpRotatorSetup.java
package lavaplayer.extensions.iprotator;

import lavaplayer.extensions.iprotator.planner.AbstractRoutePlanner;
import lavaplayer.manager.AudioPlayerManager;
import lavaplayer.source.youtube.YoutubeItemSourceManager;
import lavaplayer.source.youtube.YoutubeHttpContextFilter;
import lavaplayer.tools.http.ExtendedHttpClientBuilder;
import lavaplayer.tools.http.ExtendedHttpConfigurable;
import lavaplayer.tools.http.HttpContextFilter;
import lavaplayer.tools.http.SimpleHttpClientConnectionManager;

import java.util.ArrayList;
import java.util.List;

public class YoutubeIpRotatorSetup {
    private static final int DEFAULT_RETRY_LIMIT = 4;
    private static final HttpContextFilter DEFAULT_DELEGATE = new YoutubeHttpContextFilter();
    private static final IpRotatorRetryHandler RETRY_HANDLER = new IpRotatorRetryHandler();

    private final AbstractRoutePlanner routePlanner;
    private final List<ExtendedHttpConfigurable> mainConfiguration;
    private final List<ExtendedHttpConfigurable> searchConfiguration;
    private int retryLimit = DEFAULT_RETRY_LIMIT;
    private HttpContextFilter mainDelegate = DEFAULT_DELEGATE;
    private HttpContextFilter searchDelegate = null;

    public YoutubeIpRotatorSetup(AbstractRoutePlanner routePlanner) {
        this.routePlanner = routePlanner;
        mainConfiguration = new ArrayList<>();
        searchConfiguration = new ArrayList<>();
    }

    public YoutubeIpRotatorSetup forConfiguration(ExtendedHttpConfigurable configurable, boolean isSearch) {
        if (isSearch) {
            searchConfiguration.add(configurable);
        } else {
            mainConfiguration.add(configurable);
        }

        return this;
    }

    public YoutubeIpRotatorSetup forSource(YoutubeItemSourceManager sourceManager) {
        forConfiguration(sourceManager.getMainHttpConfiguration(), false);
        forConfiguration(sourceManager.getSearchHttpConfiguration(), true);
        forConfiguration(sourceManager.getSearchMusicHttpConfiguration(), true);
        return this;
    }

    public YoutubeIpRotatorSetup forManager(AudioPlayerManager playerManager) {
        YoutubeItemSourceManager sourceManager = playerManager.source(YoutubeItemSourceManager.class);

        if (sourceManager != null) {
            forSource(sourceManager);
        }

        return this;
    }

    public YoutubeIpRotatorSetup withRetryLimit(int retryLimit) {
        this.retryLimit = retryLimit;
        return this;
    }

    public YoutubeIpRotatorSetup withMainDelegateFilter(HttpContextFilter filter) {
        this.mainDelegate = filter;
        return this;
    }

    public YoutubeIpRotatorSetup withSearchDelegateFilter(HttpContextFilter filter) {
        this.searchDelegate = filter;
        return this;
    }

    public void setup() {
        apply(mainConfiguration, new IpRotatorFilter(mainDelegate, false, routePlanner, retryLimit));
        apply(searchConfiguration, new IpRotatorFilter(searchDelegate, true, routePlanner, retryLimit));
    }

    protected void apply(List<ExtendedHttpConfigurable> configurables, IpRotatorFilter filter) {
        for (ExtendedHttpConfigurable configurable : configurables) {
            configurable.configureBuilder(builder ->
                ((ExtendedHttpClientBuilder) builder).setConnectionManagerFactory(SimpleHttpClientConnectionManager::new)
            );

            configurable.configureBuilder(it -> {
                it.setRoutePlanner(routePlanner);
                // No retry for some exceptions we know are hopeless for retry.
                it.setRetryHandler(RETRY_HANDLER);
                // Regularly cleans up per-route connection pool which gets huge due to many routes caused by
                // each request having a unique route.
                it.evictExpiredConnections();
            });

            configurable.setHttpContextFilter(filter);
        }
    }
}

File: extensions/ip-rotator/src/main/java/lavaplayer/extensions/iprotator/YoutubeIpRotatorSetup.kt
package lavaplayer.extensions.iprotator

import lavaplayer.extensions.iprotator.planner.AbstractRoutePlanner
import lavaplayer.source.common.SourceRegistry
import lavaplayer.source.youtube.YoutubeHttpContextFilter
import lavaplayer.source.youtube.YoutubeItemSourceManager
import lavaplayer.tools.extensions.source
import lavaplayer.tools.http.ExtendedHttpConfigurable
import lavaplayer.tools.http.HttpContextFilter

class YoutubeIpRotatorSetup(routePlanner: AbstractRoutePlanner) : IpRotatorSetup(routePlanner) {
    companion object {
        private const val DEFAULT_RETRY_LIMIT = 4
        private val DEFAULT_DELEGATE: HttpContextFilter = YoutubeHttpContextFilter()
        private val RETRY_HANDLER = IpRotatorRetryHandler()

        operator fun invoke(routePlanner: AbstractRoutePlanner, build: YoutubeIpRotatorSetup.() -> Unit): YoutubeIpRotatorSetup {
            return YoutubeIpRotatorSetup(routePlanner)
                .apply(build)
        }
    }

    private val mainConfiguration = mutableListOf<ExtendedHttpConfigurable>()
    private val searchConfiguration = mutableListOf<ExtendedHttpConfigurable>()

    var retryLimit = DEFAULT_RETRY_LIMIT
    var mainDelegate = DEFAULT_DELEGATE
    var searchDelegate: HttpContextFilter? = null

    override val retryHandler: IpRotatorRetryHandler = RETRY_HANDLER

    /**
     * Applies this ip-rotator configuration to the supplied [sourceManager]
     *
     * @param sourceManager The [YoutubeItemSourceManager] to apply to.
     */
    fun applyTo(sourceManager: YoutubeItemSourceManager): YoutubeIpRotatorSetup {
        useConfiguration(sourceManager.mainHttpConfiguration, false)
        useConfiguration(sourceManager.searchHttpConfiguration, true)
        useConfiguration(sourceManager.searchMusicHttpConfiguration, true)
        return this
    }

    /**
     * Applies this ip-rotator configuration to the supplied [registry]
     *
     * @param registry The [SourceRegistry] to apply to.
     */
    fun applyTo(registry: SourceRegistry): YoutubeIpRotatorSetup {
        val sourceManager = registry.source<YoutubeItemSourceManager>()
        sourceManager?.let { applyTo(it) }
        return this
    }

    fun withRetryLimit(limit: Int): YoutubeIpRotatorSetup {
        retryLimit = limit
        return this
    }

    fun withMainDelegateFilter(filter: HttpContextFilter): YoutubeIpRotatorSetup {
        mainDelegate = filter
        return this
    }

    fun withSearchDelegateFilter(filter: HttpContextFilter?): YoutubeIpRotatorSetup {
        searchDelegate = filter
        return this
    }

    override fun setup() {
        apply(mainConfiguration, IpRotatorFilter(mainDelegate, false, routePlanner, retryLimit))
        apply(searchConfiguration, IpRotatorFilter(searchDelegate, true, routePlanner, retryLimit))
    }

    private fun useConfiguration(configurable: ExtendedHttpConfigurable, isSearch: Boolean): YoutubeIpRotatorSetup {
        val configurations = if (isSearch) searchConfiguration else mainConfiguration
        configurations.add(configurable)
        return this
    }
}
